homeassistant/components/onvif/config_flow.py

```diff
@@ -0,0 +1,349 @@
+
+    async def async_step_device(self, user_input=None):
+        """Handle WS-Discovery.
+
+        Let user choose between discovered devices and manual configuration.
+        If no device is found allow user to manually input configuration.
+        """
+        if user_input:
+
+            if CONF_MANUAL_INPUT == user_input[CONF_HOST]:
+                return await self.async_step_manual_input()
+
+            for device in self.devices:
+                name = "%s (%s:%d)" % (
+                    device[CONF_NAME],
+                    device[CONF_HOST],
+                    device[CONF_PORT],
+                )
+                if name == user_input[CONF_HOST]:
+                    self.device_id = device[CONF_DEVICE_ID]
+                    self.onvif_config = {
+                        CONF_NAME: device[CONF_NAME],
+                        CONF_HOST: device[CONF_HOST],
+                        CONF_PORT: device[CONF_PORT],
+                    }
+                    return await self.async_step_auth()
+
+        discovery = await async_discovery(self.hass)
+        for device in discovery:
+            configured = False
+            for entry in self._async_current_entries():
+                if entry.unique_id == device[CONF_DEVICE_ID]:
+                    configured = True
+                    break
+            if not configured:
+                self.devices.append(device)
+
+        LOGGER.debug("Discovered ONVIF devices %s", pformat(self.devices))
+
+        if self.devices:
+            names = []
+
+            for device in self.devices:
+                names.append(
+                    "%s (%s:%d)"
+                    % (device[CONF_NAME], device[CONF_HOST], device[CONF_PORT])
+                )
+
+            names.append(CONF_MANUAL_INPUT)
+
+            return self.async_show_form(
+                step_id="device",
+                data_schema=vol.Schema({vol.Optional(CONF_HOST): vol.In(names)}),
+            )
+
+        return await self.async_step_manual_input()
+
+    async def async_step_manual_input(self, user_input=None):
+        """Manual configuration."""
+        if user_input:
+            self.onvif_config = user_input
+            return await self.async_step_auth()
+
+        return self.async_show_form(
+            step_id="manual_input",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_NAME): str,
+                    vol.Required(CONF_HOST): str,
+                    vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
+                }
+            ),
+        )
+
+    async def async_step_auth(self, user_input=None):
+        """Username and Password configuration for ONVIF device."""
+        if user_input:
+            self.onvif_config[CONF_USERNAME] = user_input[CONF_USERNAME]
+            self.onvif_config[CONF_PASSWORD] = user_input[CONF_PASSWORD]
+            return await self.async_step_profiles()
```

Comment:
Shouldn't you verify the credentials and throw an error if it doesn't work ?

Reply:
The credentials get verified and caught in the next step. I wanted all API calls in the same place. On a credentials error it will throw it back to this step.